Component 3- **Prompts**

**What is Prompt Engineering?**

Prompt engineering refers to the process of designing and refining prompts or instructions given to a language model like ChatGPT to obtain desired outputs. It involves crafting specific and well-defined queries or requests to elicit accurate and relevant responses from the model.

The process of prompt engineering involves iterative experimentation and fine-tuning. It often requires an understanding of the models capabilities, limitations, and biases. Through trial and error, researchers can adjust prompts to achieve the desired output quality, correctness, and relevance.


**5 Best Practises for giving Prompts**

Delimeters <br>
Structured Output <br>
Few Shot Prompting <br>
Chaining the steps <br>
Show detailed solution <br>

In [ ]:
# !pip install openai
# !pip install langchain

In [ ]:
import openai

In [ ]:
openai_api_key=''

In [ ]:
from langchain.llms import OpenAI

llm = OpenAI(openai_api_key=openai_api_key, temperature = 0)

**Delimiters**

In [ ]:
text = f"""
In today's rapidly evolving world, where technology drives innovation and communication spans across continents within seconds, humanity finds itself grappling with unprecedented challenges and opportunities. The intersection of artificial intelligence, automation, and robotics promises a future where tasks once performed by humans are now delegated to machines, raising concerns about the displacement of jobs and the need to adapt our workforce. Simultaneously, advancements in medical research and biotechnology offer hope for curing diseases and extending human lifespan, but also pose ethical dilemmas regarding genetic manipulation and the boundaries of human enhancement. Meanwhile, the urgent issue of climate change demands immediate action to mitigate its catastrophic consequences, compelling nations and societies to shift towards sustainable practices and renewable energy sources. Amidst this backdrop, societal divisions fueled by political polarization and cultural differences persist, emphasizing the need for empathy, inclusivity, and dialogue to foster understanding and unity. In summary, the present era is characterized by the juxtaposition of remarkable scientific progress and complex global challenges, necessitating careful consideration, collaboration, and ethical decision-making as we navigate the uncertain terrain of the future.
"""
prompt = f"""
Summarize the text delimited by triple backticks \
into a single sentence.
```{text}```
"""
llm(prompt)

'The present era is characterized by remarkable scientific progress and complex global challenges, requiring careful consideration, collaboration, and ethical decision-making to navigate the uncertain terrain of the future.'

The present era is marked by significant scientific progress and complex global challenges, requiring careful consideration, collaboration, and ethical decision-making as we navigate the uncertain terrain of the future.


**Structured Output - JSON, HTML, XML**

In [ ]:
prompt = f"""
Generate a list of countries including USA, India and Uganda along \
with their country code and current president.
Provide them in JSON format with the following keys:
country_id, name, code, president.
"""
llm(prompt)

'\n[\n  {\n    "country_id": 1,\n    "name": "USA",\n    "code": "US",\n    "president": "Joe Biden"\n  },\n  {\n    "country_id": 2,\n    "name": "India",\n    "code": "IN",\n    "president": "Ram Nath Kovind"\n  },\n  {\n    "country_id": 3,\n    "name": "Uganda",\n    "code": "UG",\n    "president": "Yoweri Museveni"\n  }\n]'

[
  {
    "country_id": 1,
    "name": "United States of America",
    "code": "USA",
    "president": "Joe Biden"
  },
  {
    "country_id": 2,
    "name": "India",
    "code": "IND",
    "president": "Ram Nath Kovind"
  },
  {
    "country_id": 3,
    "name": "Uganda",
    "code": "UGA",
    "president": "Yoweri Museveni"
  }
]

In [ ]:
prompt = f"""
Generate a list of countries including USA, India and Uganda along \
with their country code and current president.
Provide them in HTML format with the following keys:
country_id, name, code, president.
"""
llm(prompt)

'\n<table>\n  <tr>\n    <th>country_id</th>\n    <th>name</th>\n    <th>code</th>\n    <th>president</th>\n  </tr>\n  <tr>\n    <td>1</td>\n    <td>USA</td>\n    <td>US</td>\n    <td>Joe Biden</td>\n  </tr>\n  <tr>\n    <td>2</td>\n    <td>India</td>\n    <td>IN</td>\n    <td>Ram Nath Kovind</td>\n  </tr>\n  <tr>\n    <td>3</td>\n    <td>Uganda</td>\n    <td>UG</td>\n    <td>Yoweri Museveni</td>\n  </tr>\n</table>'

<table>
  <tr>
    <th>country_id</th>
    <th>name</th>
    <th>code</th>
    <th>president</th>
  </tr>
  <tr>
    <td>1</td>
    <td>United States of America</td>
    <td>USA</td>
    <td>Joe Biden</td>
  </tr>
  <tr>
    <td>2</td>
    <td>India</td>
    <td>IND</td>
    <td>Narendra Modi</td>
  </tr>
  <tr>
    <td>3</td>
    <td>Uganda</td>
    <td>UGA</td>
    <td>Yoweri Museveni</td>
  </tr>
</table>

In [ ]:
prompt = f"""
Generate a list of countries including USA, India and Uganda along \
with their country code and current president.
Provide them in XML format with the following keys:
country_id, name, code, president.
"""
llm(prompt)

'\n<?xml version="1.0" encoding="UTF-8"?>\n<countries>\n  <country>\n    <country_id>1</country_id>\n    <name>USA</name>\n    <code>US</code>\n    <president>Joe Biden</president>\n  </country>\n  <country>\n    <country_id>2</country_id>\n    <name>India</name>\n    <code>IN</code>\n    <president>Ram Nath Kovind</president>\n  </country>\n  <country>\n    <country_id>3</country_id>\n    <name>Uganda</name>\n    <code>UG</code>\n    <president>Yoweri Museveni</president>\n  </country>\n</countries>'

<countries>
  <country>
    <country_id>1</country_id>
    <name>United States of America</name>
    <code>USA</code>
    <president>Joe Biden</president>
  </country>
  <country>
    <country_id>2</country_id>
    <name>India</name>
    <code>IND</code>
    <president>Ram Nath Kovind</president>
  </country>
  <country>
    <country_id>3</country_id>
    <name>Uganda</name>
    <code>UGA</code>
    <president>Yoweri Museveni</president>
  </country>
</countries>

**Few Shot Prompting**

In [ ]:
prompt = f"""
Your task is to answer in a funny manner

<child>: What is patience

<grandparent>: Patience is like waiting for your pizza delivery during a snowstorm when your stomach is growling louder than a hungry lion. It's that superpower that keeps you from tearing your hair out while watching the seconds crawl by on the clock. It's like trying to untangle a bunch of earphones without losing your sanity or accidentally summoning a demon.

<child>: What is generosity
"""
llm(prompt)

"\n<grandparent>: Generosity is like giving away your last slice of pizza to a friend who's had a bad day. It's like sharing your umbrella when it's pouring rain and you're already soaked. It's like giving someone a hug when they need it most."

<grandparent>: Generosity is like sharing your last cookie with your arch-nemesis, or giving away your favorite toy to a group of wild monkeys. It's that warm fuzzy feeling you get when you make someone else's day, even if it means sacrificing your own happiness. It's like being a superhero, but instead of fighting crime, you're spreading kindness and joy wherever you go.


**Specify the steps required to complete the task**

In [ ]:
text = f"""
यदा यदा हि धर्मस्य ग्लानिर्भवति भारत:।
अभ्युत्थानमधर्मस्य तदात्मानं सृजाम्यहम्॥
परित्राणाय साधूनाम् विनाशाय च दुष्कृताम्।
धर्मसंस्थापनार्थाय सम्भवामि युगे-युगे॥
"""
# example 1
prompt= f"""
Perform the following actions:
1 - Summarize the following text delimited by triple \
backticks with 1 sentence.
2 - Translate the summary into English.
3 - List each verb in the English summary.

Separate your answers with line breaks.

Text:
```{text}```
"""
llm(prompt)

"\nधर्मस्य ग्लानिर्भवति और तदात्मानं सृजाम्यहम् से भारत को परित्राण, साधूनाम् विनाश से बचाने और धर्मसंस्थापन के लिए युग-युग में सम्भव होता है।\n\nThe text speaks of India's resilience in protecting itself from destruction and upholding its Dharma through ages.\n\nVerbs: speaks, protecting, upholding.\n"

**Show Detailed Solution**

In [ ]:
prompt = f"""
Show detailed steps for the following problem
Tom bought a bicycle for $200 and spent an additional $50 on repairs. He then decided to sell it at a 20% profit. What price should Tom sell the bicycle for in order to achieve his desired profit?
"""
llm(prompt)

'\nStep 1: Calculate the total cost of the bicycle and repairs.\nTotal cost = $200 + $50 = $250\n\nStep 2: Calculate the desired profit.\nDesired profit = 20% of $250 = $50\n\nStep 3: Calculate the selling price.\nSelling price = Total cost + Desired profit = $250 + $50 = $300'

1. Calculate the total cost of the bicycle for Tom:
$200 (original cost) + $50 (repairs) = $250

2. Determine the profit percentage Tom wants to make:
20%

3. Calculate the profit amount Tom wants to make:
20% of $250 = $50

4. Add the profit amount to the total cost of the bicycle:
$250 (total cost) + $50 (profit) = $300

5. The price Tom should sell the bicycle for in order to achieve his desired profit is $300.